File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/RevokeServerViewModel.swift
import Foundation
import Observation
import CryptoKit
import Flagship
import FlagshipAPI
import FlagshipCore

/// P13 — per-server kill-switch orchestrator.
///
/// IRK-signs a `ServerRevocation` envelope (one of {lost, stolen,
/// decommissioned}) and POSTs it to .com. On success the box is dead:
/// it will refuse to boot on its next reboot. The trust gate is the
/// IRK signature alone — same level as releasing a name — combined
/// with the hold-to-confirm gesture on the sheet (see ServerDetailScreen).
@Observable
@MainActor
public final class RevokeServerViewModel {

    public enum Phase: Equatable, Sendable {
        case idle
        case signing
        case posting
        case completed
        case failed(String)
    }

    public enum Reason: String, CaseIterable, Sendable {
        case lost
        case stolen
        case decommissioned

        public var label: String {
            switch self {
            case .lost:           return "Lost"
            case .stolen:         return "Stolen"
            case .decommissioned: return "Decommissioned"
            }
        }
    }

    public private(set) var phase: Phase = .idle

    private let server: FlagshipServerClient
    private let username: () -> String?
    private let serverDomain: String
    /// Pluggable for tests: override the IRK derivation step. Default
    /// derives via `Keystore.deriveIRK(reason:)`.
    private let signer: @MainActor (String) async throws -> Curve25519.Signing.PrivateKey

    public init(
        server: FlagshipServerClient,
        serverDomain: String,
        username: @escaping () -> String?,
        signer: (@MainActor (String) async throws -> Curve25519.Signing.PrivateKey)? = nil
    ) {
        self.server = server
        self.serverDomain = serverDomain
        self.username = username
        self.signer = signer ?? { reason in
            try await Keystore.deriveIRK(reason: reason)
        }
    }

    public func run(reason: Reason) async {
        guard let user = username(), !user.isEmpty else {
            phase = .failed("No active account on this device.")
            return
        }
        phase = .signing
        let irk: Curve25519.Signing.PrivateKey
        do {
            irk = try await signer("Revoke server \(serverDomain)")
        } catch {
            phase = .failed("Couldn't access your account key: \(error.localizedDescription)")
            return
        }
        let issuedAt = Int64(Date().timeIntervalSince1970 * 1000)
        let canonical = ServerRevocationClaim.canonicalBytes(
            userId: user,
            revokedServerId: serverDomain,
            reason: reason.rawValue,
            issuedAt: issuedAt
        )
        let signature: Data
        do {
            signature = try irk.signature(for: canonical)
        } catch {
            phase = .failed("Couldn't sign: \(error.localizedDescription)")
            return
        }

        phase = .posting
        do {
            try await server.revokeServer(
                ServerRevocationRequest(
                    request: .init(
                        userId: user,
                        revokedServerId: serverDomain,
                        reason: reason.rawValue,
                        issuedAt: issuedAt
                    ),
                    signature: HexUtil.encode(signature)
                )
            )
        } catch ScreensClientError.http(let status, _) where status == 403 {
            phase = .failed("The server rejected the request. Sign in again and retry.")
            return
        } catch ScreensClientError.http(let status, _) where status == 404 {
            phase = .failed("That server is already gone — nothing to revoke.")
            return
        } catch let error as ScreensClientError {
            // UX-B — plain language, no raw status code or server message;
            // UX-A — a cert-pin mismatch reads as "someone may be intercepting".
            phase = .failed(error.errorDescription ?? "That didn't work. Try again in a moment.")
            return
        } catch {
            phase = .failed("Couldn't reach the server. Check your connection and try again.")
            return
        }
        phase = .completed
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/RevokeServerViewModel.kt
// P13 — Kotlin mirror of FlagshipUI's RevokeServerViewModel. Signs +
// POSTs the per-server kill-switch envelope. See the Swift source for
// the canonical commentary.

package com.flagshipserver.app.viewmodels

import androidx.lifecycle.ViewModel
import com.flagshipserver.app.api.FlagshipServerClient
import com.flagshipserver.app.api.ServerRevocationRequest
import com.flagshipserver.app.core.HexUtil
import com.flagshipserver.app.core.HttpException
import com.flagshipserver.app.core.ServerRevocationClaim
import com.flagshipserver.app.keystore.Keystore
import com.google.crypto.tink.subtle.Ed25519Sign
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

sealed interface RevokeServerPhase {
    data object Idle : RevokeServerPhase
    data object Signing : RevokeServerPhase
    data object Posting : RevokeServerPhase
    data object Completed : RevokeServerPhase
    data class Failed(val message: String) : RevokeServerPhase
}

enum class RevokeServerReason(val wire: String, val label: String) {
    LOST("lost", "Lost"),
    STOLEN("stolen", "Stolen"),
    DECOMMISSIONED("decommissioned", "Decommissioned");

    companion object {
        fun fromWire(s: String): RevokeServerReason? = entries.firstOrNull { it.wire == s }
    }
}

class RevokeServerViewModel(
    private val server: FlagshipServerClient,
    private val serverDomain: String,
    private val username: () -> String?,
    /** Pluggable for tests. Default uses the real Keystore-backed IRK. */
    // Slice D — releasing/revoking a server name is SENSITIVE (serverRevoke.ts
    // gates it on the admin master root): sign with the admin root when held,
    // else the owner IRK (legacy). Canonical bytes unchanged.
    private val signer: suspend (reason: String) -> Ed25519Sign = { r -> Keystore.adminSigningKey(r) },
    private val now: () -> Long = { System.currentTimeMillis() },
) : ViewModel() {
    private val _phase = MutableStateFlow<RevokeServerPhase>(RevokeServerPhase.Idle)
    val phase: StateFlow<RevokeServerPhase> = _phase.asStateFlow()

    suspend fun run(reason: RevokeServerReason) {
        val user = username()
        if (user.isNullOrEmpty()) {
            _phase.value = RevokeServerPhase.Failed("No active account on this device.")
            return
        }
        _phase.value = RevokeServerPhase.Signing
        val irk: Ed25519Sign
        try {
            irk = signer("Revoke server $serverDomain")
        } catch (e: Throwable) {
            _phase.value = RevokeServerPhase.Failed("Couldn't access your account key: ${e.message}")
            return
        }
        val issuedAt = now()
        val canonical = ServerRevocationClaim.canonicalBytes(
            userId = user,
            revokedServerId = serverDomain,
            reason = reason.wire,
            issuedAt = issuedAt,
        )
        val signature: ByteArray
        try {
            signature = irk.sign(canonical)
        } catch (e: Throwable) {
            _phase.value = RevokeServerPhase.Failed("Couldn't sign: ${e.message}")
            return
        }

        _phase.value = RevokeServerPhase.Posting
        try {
            server.revokeServer(
                ServerRevocationRequest(
                    request = ServerRevocationRequest.Inner(
                        userId = user,
                        revokedServerId = serverDomain,
                        reason = reason.wire,
                        issuedAt = issuedAt,
                    ),
                    signature = HexUtil.encode(signature),
                ),
            )
        } catch (e: HttpException) {
            val friendly = when (e.status) {
                403 -> "The server rejected the request. Sign in again and retry."
                404 -> "That server is already gone — nothing to revoke."
                else -> "Server error (${e.status}): ${e.body}"
            }
            _phase.value = RevokeServerPhase.Failed(friendly)
            return
        } catch (e: Throwable) {
            _phase.value = RevokeServerPhase.Failed("Couldn't reach the server: ${e.message}")
            return
        }
        _phase.value = RevokeServerPhase.Completed
    }
}
